File: app/src/main/java/com/stedi/multitouchpaint/MainActivity.java
package com.stedi.multitouchpaint;

import android.Manifest;
import android.app.Activity;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.os.Bundle;
import android.provider.MediaStore;
import android.support.v4.app.ActivityCompat;
import android.support.v4.content.ContextCompat;

import com.squareup.otto.Subscribe;
import com.stedi.multitouchpaint.background.BitmapGetter;
import com.stedi.multitouchpaint.background.BitmapSaver;
import com.stedi.multitouchpaint.background.PendingRunnables;
import com.stedi.multitouchpaint.dialogs.BrushColorDialog;
import com.stedi.multitouchpaint.dialogs.BrushThicknessDialog;
import com.stedi.multitouchpaint.dialogs.ExitDialog;
import com.stedi.multitouchpaint.dialogs.FileWorkDialog;
import com.stedi.multitouchpaint.history.Brush;
import com.stedi.multitouchpaint.view.CanvasView;
import com.stedi.multitouchpaint.view.WorkPanel;

import butterknife.BindView;
import butterknife.ButterKnife;

public class MainActivity extends Activity {
    private final String KEY_BRUSH = "key_brush";
    private final int REQUEST_GET_IMAGE = 111;
    private final int REQUEST_PERM_READ = 222;
    private final int REQUEST_PERM_WRITE = 333;

    @BindView(R.id.main_activity_canvas_view)
    CanvasView canvasView;
    @BindView(R.id.main_activity_work_panel)
    WorkPanel workPanel;

    private Brush brush;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        App.getBus().register(this);

        setContentView(R.layout.main_activity);
        ButterKnife.bind(this);

        if (savedInstanceState != null)
            brush = savedInstanceState.getParcelable(KEY_BRUSH);
        if (brush == null)
            brush = Brush.createDefault();

        canvasView.setBrush(brush);
        workPanel.setBrush(brush);
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putParcelable(KEY_BRUSH, brush);
    }

    @Override
    protected void onResume() {
        super.onResume();
        PendingRunnables.getInstance().onResume();
    }

    @Override
    protected void onPause() {
        super.onPause();
        PendingRunnables.getInstance().onPause();
    }

    @Override
    protected void onDestroy() {
        App.getBus().unregister(this);
        super.onDestroy();
    }

    @Override
    public void onBackPressed() {
        if (canvasView.inPipetteMode()) {
            canvasView.disablePipette();
            workPanel.show();
            return;
        }
        if (workPanel.isShown())
            workPanel.hide();
        else
            workPanel.show();
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, String permissions[], int[] grantResults) {
        if (grantResults.length == 0 || grantResults[0] != PackageManager.PERMISSION_GRANTED)
            return;
        switch (requestCode) {
            case REQUEST_PERM_WRITE:
                saveCanvasViewImage();
                break;
            case REQUEST_PERM_READ:
                startPickImageIntent();
                break;
            default:
                break;
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (resultCode == RESULT_OK && requestCode == REQUEST_GET_IMAGE) {
            new BitmapGetter(data.getData(), canvasView.getWidth(), canvasView.getHeight()).start();
        }
    }

    @Subscribe
    public void onBrushUpdate(Brush brush) {
        this.brush = brush;
        canvasView.setBrush(brush);
        workPanel.setBrush(brush);
    }

    @Subscribe
    public void onWorkPanelEvent(WorkPanel.Callback callback) {
        if (canvasView.isDrawing())
            return;
        switch (callback) {
            case ON_FILE_WORK_CLICK:
                new FileWorkDialog().show(getFragmentManager(), FileWorkDialog.class.getName());
                break;
            case ON_PIPETTE_CLICK:
                canvasView.activatePipette();
                workPanel.hide();
                break;
            case ON_COLOR_CLICK:
                BrushColorDialog.newInstance(brush).show(getFragmentManager(), BrushColorDialog.class.getName());
                break;
            case ON_THICKNESS_CLICK:
                BrushThicknessDialog.newInstance(brush).show(getFragmentManager(), BrushThicknessDialog.class.getName());
                break;
            case ON_UNDO_CLICK:
                canvasView.undo();
                break;
            case ON_EXIT_CLICK:
                new ExitDialog().show(getFragmentManager(), ExitDialog.class.getName());
                break;
            default:
                break;
        }
    }

    @Subscribe
    public void onFileWorkDialogEvent(FileWorkDialog.Callback callback) {
        switch (callback) {
            case ON_NEW_FILE:
                canvasView.clearPicture();
                break;
            case ON_OPEN:
                if (checkForPermission(Manifest.permission.READ_EXTERNAL_STORAGE, REQUEST_PERM_READ))
                    startPickImageIntent();
                break;
            case ON_SAVE:
                if (checkForPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE, REQUEST_PERM_WRITE))
                    saveCanvasViewImage();
                break;
            default:
                break;
        }
    }

    @Subscribe
    public void onExitDialogEvent(ExitDialog.Callback callback) {
        finish();
    }

    @Subscribe
    public void onBitmapSaverEvent(BitmapSaver.Callback callback) {
        switch (callback) {
            case BITMAP_SAVED:
                App.showToast(R.string.image_successfully_saved);
                break;
            case FAILED_TO_SAVE:
                App.showToast(R.string.failed_to_save_image);
                break;
            case CANT_SAVE:
                App.showToast(R.string.cant_save_image);
                break;
            default:
                break;
        }
    }

    @Subscribe
    public void onGalleryBitmapGetterEvent(BitmapGetter.Callback callback) {
        Bitmap bitmap = callback.bitmap;
        if (bitmap != null)
            canvasView.setPicture(bitmap);
        else
            App.showToast(R.string.failed_to_load_image);
    }

    private void saveCanvasViewImage() {
        new BitmapSaver(canvasView.generatePicture()).start();
    }

    private void startPickImageIntent() {
        Intent intent = new Intent(Intent.ACTION_PICK, MediaStore.Images.Media.EXTERNAL_CONTENT_URI);
        startActivityForResult(intent, REQUEST_GET_IMAGE);
    }

    private boolean checkForPermission(String permission, int requestCode) {
        if (ContextCompat.checkSelfPermission(this, permission) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this, new String[]{permission}, requestCode);
            return false;
        }
        return true;
    }
}

File: app/src/main/java/com/stedi/multitouchpaint/MainActivity.kt
package com.stedi.multitouchpaint

import android.Manifest
import android.app.Activity
import android.content.ActivityNotFoundException
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import android.provider.MediaStore
import android.support.v4.app.ActivityCompat
import android.support.v4.content.ContextCompat
import butterknife.BindView
import butterknife.ButterKnife
import com.squareup.otto.Subscribe
import com.stedi.multitouchpaint.background.BitmapGetter
import com.stedi.multitouchpaint.background.BitmapSaver
import com.stedi.multitouchpaint.background.PendingRunnables
import com.stedi.multitouchpaint.data.Brush
import com.stedi.multitouchpaint.dialogs.*
import com.stedi.multitouchpaint.painters.PathPainter
import com.stedi.multitouchpaint.painters.PipettePainter
import com.stedi.multitouchpaint.view.CanvasView
import com.stedi.multitouchpaint.view.WorkPanel

class MainActivity : Activity() {
    private val KEY_BRUSH = "KEY_BRUSH"
    private val REQUEST_GET_IMAGE = 111
    private val REQUEST_PERM_READ = 222
    private val REQUEST_PERM_WRITE = 333

    @BindView(R.id.main_activity_canvas_view)
    lateinit var canvasView: CanvasView

    @BindView(R.id.main_activity_work_panel)
    lateinit var workPanel: WorkPanel

    private lateinit var brush: Brush

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        App.getBus().register(this)
        setContentView(R.layout.main_activity)
        ButterKnife.bind(this)
        brush = savedInstanceState?.getParcelable(KEY_BRUSH) ?: Brush.createDefault()
        canvasView.setBrush(brush)
        canvasView.setPainter(PathPainter.getInstance())
        workPanel.setBrush(brush)
    }

    override fun onResume() {
        super.onResume()
        PendingRunnables.getInstance().onResume()
    }

    override fun onPause() {
        super.onPause()
        PendingRunnables.getInstance().onPause()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putParcelable(KEY_BRUSH, brush)
    }

    override fun onDestroy() {
        super.onDestroy()
        App.getBus().unregister(this)
    }

    override fun onBackPressed() {
        if (onPipettePainterBackPressed()) {
            return
        }
        if (workPanel.isShown()) {
            workPanel.hide()
        } else {
            workPanel.show()
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray) {
        if (grantResults.isEmpty() || grantResults[0] != PackageManager.PERMISSION_GRANTED) {
            return
        }
        when (requestCode) {
            REQUEST_PERM_WRITE -> saveCanvasViewImage()
            REQUEST_PERM_READ -> startPickImageIntent()
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent) {
        if (resultCode == RESULT_OK && requestCode == REQUEST_GET_IMAGE) {
            WaitDialog.start(fragmentManager)
            BitmapGetter(data.data, canvasView.width, canvasView.height).start()
        }
    }

    @Subscribe
    fun onBrushUpdate(brush: Brush) {
        this.brush = brush
        canvasView.setBrush(brush)
        workPanel.setBrush(brush)
    }

    @Subscribe
    fun onWorkPanelEvent(callback: WorkPanel.Callback) {
        if (canvasView.isDrawing()) {
            return
        }
        when (callback) {
            WorkPanel.Callback.ON_FILE_WORK_CLICK -> FileWorkDialog().show(fragmentManager, FileWorkDialog::class.java.name)
            WorkPanel.Callback.ON_PIPETTE_CLICK -> {
                canvasView.setPainter(PipettePainter(canvasView.generatePicture()))
                workPanel.hide()
            }
            WorkPanel.Callback.ON_COLOR_CLICK -> BrushColorDialog.newInstance(brush).show(fragmentManager, BrushColorDialog::class.java.name)
            WorkPanel.Callback.ON_THICKNESS_CLICK -> BrushThicknessDialog.newInstance(brush).show(fragmentManager, BrushThicknessDialog::class.java.name)
            WorkPanel.Callback.ON_UNDO_CLICK -> canvasView.undo()
            WorkPanel.Callback.ON_EXIT_CLICK -> ExitDialog().show(fragmentManager, ExitDialog::class.java.name)
        }
    }

    @Subscribe
    fun onFileWorkDialogEvent(callback: FileWorkDialog.Callback) {
        when (callback) {
            FileWorkDialog.Callback.ON_NEW_FILE -> canvasView.clearPicture()
            FileWorkDialog.Callback.ON_OPEN -> {
                if (checkForPermission(Manifest.permission.READ_EXTERNAL_STORAGE, REQUEST_PERM_READ)) {
                    startPickImageIntent()
                }
            }
            FileWorkDialog.Callback.ON_SAVE -> {
                if (checkForPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE, REQUEST_PERM_WRITE)) {
                    saveCanvasViewImage()
                }
            }
        }
    }

    @Subscribe
    fun onExitDialogEvent(callback: ExitDialog.Callback) {
        finish()
    }

    @Subscribe
    fun onBitmapSaverEvent(callback: BitmapSaver.Callback) {
        WaitDialog.stop(fragmentManager)
        when (callback) {
            BitmapSaver.Callback.BITMAP_SAVED -> App.showToast(R.string.image_successfully_saved)
            BitmapSaver.Callback.FAILED_TO_SAVE -> App.showToast(R.string.failed_to_save_image)
            BitmapSaver.Callback.CANT_SAVE -> App.showToast(R.string.cant_save_image)
        }
    }

    @Subscribe
    fun onGalleryBitmapGetterEvent(callback: BitmapGetter.Callback) {
        WaitDialog.stop(fragmentManager)
        if (callback.bitmap != null) {
            canvasView.setPicture(callback.bitmap)
        } else {
            App.showToast(R.string.failed_to_load_image)
        }
    }

    private fun saveCanvasViewImage() {
        WaitDialog.start(fragmentManager)
        BitmapSaver(canvasView.generatePicture()).start()
    }

    private fun startPickImageIntent() {
        try {
            startActivityForResult(Intent(Intent.ACTION_PICK, MediaStore.Images.Media.EXTERNAL_CONTENT_URI), REQUEST_GET_IMAGE)
        } catch (e: ActivityNotFoundException) {
            e.printStackTrace()
            App.showToast(R.string.unknown_error)
        }
    }

    private fun checkForPermission(permission: String, requestCode: Int): Boolean {
        if (ContextCompat.checkSelfPermission(this, permission) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this, arrayOf(permission), requestCode)
            return false
        }
        return true
    }

    private fun onPipettePainterBackPressed(): Boolean {
        val painter = canvasView.getPainter()
        if (painter is PipettePainter) {
            if (!painter.isDrawing()) {
                val newBrush = Brush.copy(brush)
                newBrush.setColor(painter.getColor())
                App.getBus().post(newBrush)
                canvasView.setPainter(PathPainter.getInstance())
                workPanel.show()
            }
            return true
        }
        return false
    }
}